community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -54,21 +54,91 @@
 TOT_REQ_CNT = 1000
 _MAINTENANCE_SBATCH_SCRIPT_PATH = dirs.custom_scripts / "perform_maintenance.sh"
 
-NodeStatus = Enum(
-    "NodeStatus",
-    (
-        "orphan",
-        "power_down",
-        "preempted",
-        "restore",
-        "resume",
-        "terminated",
-        "unbacked",
-        "unchanged",
-        "unknown",
-    ),
-)
+class NodeAction(Protocol):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        ...
+
+    def __hash__(self):
+        return hash(type(self))
+
+    def __eq__(self, other):
+        return type(self) == type(other)
+
+class NodeActionResume(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """resume nodes via scontrol"""
+        log.info(f"{count} instances to resume ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=power_up")
+
+class NodeActionIdle(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """idle nodes"""
+        log.info(f"{count} nodes to idle ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=resume")
+
+class NodeActionRestart(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """start instances for nodes"""
+        log.info(f"{count} instances restarted ({hostlist})")
+        start_instances(nodes)
+
+class NodeActionDown(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int, reason:str="Instance stopped/deleted") -> None:
```

Comment:
`reason` should be a parameter of the the `NodeActionDown` instance, not parameter of `apply`.